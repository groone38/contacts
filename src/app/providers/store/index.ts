import { configureStore } from "@reduxjs/toolkit";
import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
import { auth, contacts } from "src/features";

export const store = configureStore({
  reducer: {
    auth,
    contacts,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
export const useAppDispatch = () => useDispatch<typeof store.dispatch>();
