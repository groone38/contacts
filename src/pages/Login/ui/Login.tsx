import React from "react";
import { useForm, SubmitHandler } from "react-hook-form";
import { Input } from "src/shared/Input";
import { useAppDispatch, useAppSelector } from "src/app/providers/store";
import { signInUser } from "src/features/model/reducers/AuthSlice";
import { useNavigate } from "react-router-dom";
import { Loader } from "src/shared/ui/Loader";

import { Button } from "src/shared/ui/Button";
import { LinkButton } from "src/features/ui/LinkButton";

import classes from "./Login.module.scss";
interface LoginData {
  email: string;
  password: string;
}

const Login = () => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<LoginData>({
    mode: "all",
    defaultValues: {
      email: "",
      password: "",
    },
  });
  const dispatch = useAppDispatch();
  const loading = useAppSelector((state) => state.auth.loading);
  const navigate = useNavigate();
  const onSubmit: SubmitHandler<LoginData> = (data) => {
    dispatch(signInUser(data)).then(() => navigate("/"));
  };

  return (
    <div className={classes.wrap}>
      {loading && <Loader />}
      <form onSubmit={handleSubmit(onSubmit)} className={classes.form}>
        <h2 className={classes.form__title}>Sing in</h2>
        <Input
          id="email"
          name="email"
          placeholder="Enter you email"
          register={register}
          required="Поле Email не может быть пустым!"
          type="email"
          label="Email"
          errors={errors?.email?.message}
        />
        <Input
          id="password"
          name="password"
          placeholder="Enter you password"
          register={register}
          required="Поле Password не может быть пустым!"
          type="password"
          label="Password"
          errors={errors?.password?.message}
        />
        <div className={classes.form__btn}>
          <Button type="submit">Sing in</Button>
          <LinkButton link={"/register"}>Sing Up</LinkButton>
        </div>
      </form>
    </div>
  );
};

export default Login;
