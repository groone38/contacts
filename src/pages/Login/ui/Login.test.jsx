import { act, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import Login from "./Login";
import { Provider } from "react-redux";
import { store } from "../../../app/providers/store/index";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

describe("Login", () => {
  it("render component", async () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </BrowserRouter>
    );
    expect(screen.getByTestId("title")).toBeInTheDocument();
  });
  it("validation email and password error", async () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </BrowserRouter>
    );

    await act(() => {
      userEvent.click(getByTestId("btn"));
    });

    expect(
      screen.getByText("Поле Password не может быть пустым!")
    ).toBeInTheDocument();

    expect(
      screen.getByText("Поле Email не может быть пустым!")
    ).toBeInTheDocument();
  });
  it("validation email error", async () => {
    const { getByTestId, getByLabelText } = render(
      <BrowserRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </BrowserRouter>
    );
    await act(() => {
      userEvent.type(getByLabelText("Email:"), "groone38@yandex.ru");
    });

    await act(() => {
      userEvent.click(getByTestId("btn"));
    });
    screen.debug();
    expect(
      screen.getByText("Поле Password не может быть пустым!")
    ).toBeInTheDocument();
  });

  it("validation password error", async () => {
    const { getByTestId, getByLabelText } = render(
      <BrowserRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </BrowserRouter>
    );
    await act(() => {
      userEvent.type(getByLabelText("Password:"), "groone38@yandex.ru");
    });

    await act(() => {
      userEvent.click(getByTestId("btn"));
    });
    screen.debug();
    expect(
      screen.getByText("Поле Email не может быть пустым!")
    ).toBeInTheDocument();
  });

  it("not validation errors", async () => {
    const { getByTestId, getByLabelText } = render(
      <BrowserRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </BrowserRouter>
    );
    await act(() => {
      userEvent.type(getByLabelText("Email:"), "groone38@yandex.ru");
    });
    await act(() => {
      userEvent.type(getByLabelText("Password:"), "groone38@yandex.ru");
    });
    await act(() => {
      userEvent.click(getByTestId("btn"));
    });
    screen.debug();
    expect(screen.queryByText("Поле Email не может быть пустым!")).toBeNull();
    expect(
      screen.queryByText("Поле Password не может быть пустым!")
    ).toBeNull();
  });
  it("Login snapchot", () => {
    const login = render(
      <BrowserRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </BrowserRouter>
    );
    expect(login).toMatchSnapshot();
  });
});
