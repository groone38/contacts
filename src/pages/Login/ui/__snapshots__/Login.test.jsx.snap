// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Login snapchot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="wrap"
      >
        <div
          class="overlay"
        >
          <div
            class="lds_spinner"
          >
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <form
          class="form"
        >
          <h2
            class="form__title"
            data-testid="title"
          >
            Sing in
          </h2>
          <div
            class="input"
          >
            <label
              for="email"
            >
              Email
              : 
            </label>
            <input
              id="email"
              name="email"
              placeholder="Enter you email"
              type="email"
            />
          </div>
          <div
            class="input"
          >
            <label
              for="password"
            >
              Password
              : 
            </label>
            <input
              id="password"
              name="password"
              placeholder="Enter you password"
              type="password"
            />
          </div>
          <div
            class="form__btn"
          >
            <button
              class="btn"
              data-testid="btn"
              type="submit"
            >
              Sing in
            </button>
            <a
              class="link"
              href="/register"
            >
              Sing Up
            </a>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="wrap"
    >
      <div
        class="overlay"
      >
        <div
          class="lds_spinner"
        >
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
      <form
        class="form"
      >
        <h2
          class="form__title"
          data-testid="title"
        >
          Sing in
        </h2>
        <div
          class="input"
        >
          <label
            for="email"
          >
            Email
            : 
          </label>
          <input
            id="email"
            name="email"
            placeholder="Enter you email"
            type="email"
          />
        </div>
        <div
          class="input"
        >
          <label
            for="password"
          >
            Password
            : 
          </label>
          <input
            id="password"
            name="password"
            placeholder="Enter you password"
            type="password"
          />
        </div>
        <div
          class="form__btn"
        >
          <button
            class="btn"
            data-testid="btn"
            type="submit"
          >
            Sing in
          </button>
          <a
            class="link"
            href="/register"
          >
            Sing Up
          </a>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
